File: web/src/stores/cors.ts
// Store for the read-only per-bucket CORS panel: whether the panel is open, the
// current bucket's rules (or null = "no CORS configured"), and the load mutator.
// Display only — there is no add/edit/delete here; management is the real S3 API
// (PutBucketCors/DeleteBucketCors), the fidelity point. IO goes through the
// read-only `/_/api/buckets/{bucket}/cors` seam.

import { signal } from "zero";
import type { CorsInfo } from "../lib/api.ts";
import { getCors } from "../lib/api.ts";

/** Whether the CORS panel is showing (vs. the folder/search view). */
export const panelOpen = signal<boolean>(false);
/** The current bucket's rules, or `null` when it has no CORS configured. */
export const rules = signal<CorsInfo[] | null>(null);
/** The bucket whose rules are currently loaded. */
export const loadedBucket = signal<string | null>(null);

/**
 * Open the CORS panel for `bucket` and load its rules.
 * @param {string} bucket
 * @returns {Promise<void>}
 */
export async function openPanel(bucket: string): Promise<void> {
  panelOpen.set(true);
  await load(bucket);
}

/** Close the CORS panel (back to the folder/search view). */
export function closePanel(): void {
  panelOpen.set(false);
}

/**
 * Load the CORS rules for `bucket` into the panel. A bucket with no config sets
 * `rules` to `null` (the empty state).
 * @param {string} bucket
 * @returns {Promise<void>}
 */
export async function load(bucket: string): Promise<void> {
  loadedBucket.set(bucket);
  const res = await getCors(bucket);
  rules.set(res.cors);
}

File: src/api/cors.rs
//! Read-only per-bucket CORS display under `GET /_/api/buckets/{bucket}/cors`.
//!
//! This seam makes a bucket's CORS config **visible** in the web UI without
//! duplicating the management surface: creating/editing/deleting CORS stays the
//! real S3 API (`PutBucketCors`/`DeleteBucketCors`) — the fidelity point. It
//! reads the same `bucket_cors` SQLite table the enforcement path does, so the
//! display reflects the live config with no restart. Any non-GET method is
//! `405` (there is no write endpoint here, by design).

use std::sync::Arc;

use hyper::{Method, Response, StatusCode};
use s3s::Body;
use serde::Serialize;

use crate::http::AppState;

/// Dispatch by method: `GET` returns the bucket's rules (or `null`); anything
/// else is `405` (management is the S3 API, not this seam).
pub fn route(method: Method, state: &Arc<AppState>, bucket: String) -> Response<Body> {
    match method {
        Method::GET => get(state, &bucket),
        m => super::error(
            StatusCode::METHOD_NOT_ALLOWED,
            "MethodNotAllowed",
            format!("{m} not allowed on cors (management is the S3 API)"),
        ),
    }
}

/// `{ "cors": [...] | null }` — the bucket's rules, or `null` when it has none
/// (the "no CORS configured" empty state the UI renders).
#[derive(Serialize)]
struct CorsResponse {
    cors: Option<Vec<crate::cors::CorsRule>>,
}

/// `GET …/cors` — read the stored rules from the same table the enforcement path
/// uses.
fn get(state: &Arc<AppState>, bucket: &str) -> Response<Body> {
    match state.db.get_bucket_cors(bucket) {
        Ok(Some(json)) => match crate::cors::parse_rules(&json) {
            Ok(rules) => super::json(StatusCode::OK, &CorsResponse { cors: Some(rules) }),
            Err(e) => super::error(
                StatusCode::INTERNAL_SERVER_ERROR,
                "InternalError",
                e.to_string(),
            ),
        },
        Ok(None) => super::json(StatusCode::OK, &CorsResponse { cors: None }),
        Err(e) => super::error(
            StatusCode::INTERNAL_SERVER_ERROR,
            "InternalError",
            e.to_string(),
        ),
    }
}
